services/streammanager/googlesheets/googlesheetsmanager.go

```diff
@@ -0,0 +1,236 @@
+		PrivateKey: []byte(credentialsFile.PrivateKey),
+		Scopes: []string{
+			"https://www.googleapis.com/auth/spreadsheets",
+		},
+		TokenURL: tokenURI,
+	}
+	googleAPIService.Jwt = jwtconfig
+	service, err := GenerateServiceWithrefreshToken(*jwtconfig)
+	googleAPIService.Service = service
+	return &googleAPIService, err
+}
+
+func Produce(jsonData json.RawMessage, producer interface{}, destConfig interface{}) (statusCode int, respStatus string, responseMessage string) {
+
+	// Here we recieve the transformed json with all the required mappings
+	parsedJSON := gjson.ParseBytes(jsonData)
+	// Storing sheet_id and sheet_tab for the specific message
+	spreadSheetId := parsedJSON.Get("spreadSheetId").String()
+	spreadSheetTab := parsedJSON.Get("spreadSheetTab").String()
+
+	var pServiceMessage string
+	// First we are pinging to google-sheets using existing client
+	pkgLogger.Debug("[Google Sheets] Pingging to Sheets using existing client", googleAPIService)
+	_, pingErr := googleAPIService.Service.Spreadsheets.Get(spreadSheetId).Do()
+	// If we encounter any error we are handling that error and parsing it into pServiceMessage
+	if pingErr != nil {
+		pkgLogger.Debug("[Google Sheets] Ping Error", pingErr.Error())
+		_, pServiceMessage = HandleServiceError(pingErr)
+
+	}
+	var serviceErr error
+	// With the formatted error message we are interested in checking if the error string has
+	// the following term in it, if so we want to update our client with new access token
+	if strings.Contains(pServiceMessage, "token expired and refresh token is not set") {
+		pkgLogger.Debug("[Google Sheets] Generating New Client")
+		// Here we are updating the client with new generated client
+		googleAPIService.Service, serviceErr = GenerateServiceWithrefreshToken(*googleAPIService.Jwt)
+		// If we face error while generating the client we lodge error with status code 400
+		if serviceErr != nil {
+			pkgLogger.Debug("[Google Sheets] Failed to Generate New Client")
+			respStatus = "Faliure"
+			responseMessage = serviceErr.Error()
+			return 400, respStatus, responseMessage
+		}
+	}
+
+	// With our global - sheet client now updated we are ready to proceed for further
+	// api calls.
+
+	// Here we extract the keys, values of the transformed json into two string arrays
+	// because google sheets-api works with array of data if any error occurs here we
+	// log it as parse error for that event
+	keys, values, parseErr := ParseTransformedData(parsedJSON)
+
+	if parseErr != nil {
+		respStatus = "Faliure"
+		responseMessage = "[GoogleSheets] error :: Failed to parse transformed data ::" + parseErr.Error()
+		return 400, respStatus, responseMessage
+
+	}
+	// Converting the string array of keys to an array of interface compatibale with sheets-api (Keys relates to sheet header)
+	headerRow := GetSheetsData(keys)
+	// Creating value range for inserting row into sheet
+	var vr sheets.ValueRange
+	vr.MajorDimension = "ROWS"
+	vr.Range = spreadSheetTab + "!A1"
```

Comment:
What is this for?

Reply:
This is a specific `struct` exposed by google-sheets-api for marking which way we want to add the data. ROW or COLUMN along with from which Starting point in Particular Sheet. This is required for using the `sheets-service` APIs